File: backend/src/main/java/com/unbora/api/domain/adminauth/AdminAuthService.java
package com.unbora.api.domain.adminauth;

import com.unbora.api.common.PasswordUtil;
import com.unbora.api.common.exception.ApiException;
import com.unbora.api.domain.adminauth.dto.*;
import com.unbora.api.domain.user.PortalUser;
import com.unbora.api.domain.user.PortalUserRepository;
import com.unbora.api.security.JwtService;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;
import java.util.List;
import java.util.UUID;

@Service
public class AdminAuthService {

    private final PortalUserRepository portalUserRepository;
    private final JwtService jwtService;
    private final String superadminEmail;
    private final String superadminPassword;
    private final String superadminName;

    public AdminAuthService(
            PortalUserRepository portalUserRepository,
            JwtService jwtService,
            @Value("${unbora.superadmin.email:}") String superadminEmail,
            @Value("${unbora.superadmin.password:}") String superadminPassword,
            @Value("${unbora.superadmin.name:Super Admin}") String superadminName
    ) {
        this.portalUserRepository = portalUserRepository;
        this.jwtService = jwtService;
        this.superadminEmail = superadminEmail != null ? superadminEmail.trim().toLowerCase() : "";
        this.superadminPassword = superadminPassword != null ? superadminPassword.trim() : "";
        this.superadminName = superadminName != null ? superadminName.trim() : "Super Admin";
    }

    public AdminLoginResponseDto login(AdminLoginDto dto) {
        String email = dto.email().trim().toLowerCase();

        // 1. Check superadmin environment credentials
        if (!superadminEmail.isBlank() && !superadminPassword.isBlank() && email.equals(superadminEmail)) {
            if (!dto.password().equals(superadminPassword)) {
                throw new ApiException(HttpStatus.UNAUTHORIZED, "E-mail ou senha inválidos.");
            }

            AdminSessionDto session = new AdminSessionDto(
                    "superadmin",
                    superadminName,
                    superadminEmail,
                    "superadmin"
            );
            String token = jwtService.generateToken(session.id(), session.email(), session.name(), session.role());
            return new AdminLoginResponseDto(token, session);
        }

        // 2. Check portal users in database
        PortalUser portalUser = portalUserRepository.findByEmail(email)
                .orElseThrow(() -> new ApiException(HttpStatus.UNAUTHORIZED, "E-mail ou senha inválidos."));

        if (!PasswordUtil.verify(dto.password(), portalUser.getPasswordHash())) {
            throw new ApiException(HttpStatus.UNAUTHORIZED, "E-mail ou senha inválidos.");
        }

        AdminSessionDto session = new AdminSessionDto(
                portalUser.getId(),
                portalUser.getName(),
                portalUser.getEmail(),
                portalUser.getRole()
        );
        String token = jwtService.generateToken(session.id(), session.email(), session.name(), session.role());
        return new AdminLoginResponseDto(token, session);
    }

    public List<PortalUserResponseDto> listPortalUsers() {
        return portalUserRepository.findAllByOrderByCreatedAtDesc().stream()
                .map(u -> new PortalUserResponseDto(
                        u.getId(),
                        u.getName(),
                        u.getEmail(),
                        u.getRole(),
                        u.getCreatedAt() != null ? u.getCreatedAt().toString() : ""
                ))
                .toList();
    }

    @Transactional
    public PortalUserResponseDto createPortalUser(CreatePortalUserDto dto) {
        String email = dto.email().trim().toLowerCase();

        if (!superadminEmail.isBlank() && email.equals(superadminEmail)) {
            throw new ApiException(HttpStatus.CONFLICT, "Este e-mail já está reservado ao superadmin.");
        }

        if (portalUserRepository.findByEmail(email).isPresent()) {
            throw new ApiException(HttpStatus.CONFLICT, "Este e-mail já está cadastrado no portal.");
        }

        PortalUser portalUser = new PortalUser();
        portalUser.setId(UUID.randomUUID().toString());
        portalUser.setName(dto.name().trim());
        portalUser.setEmail(email);
        portalUser.setPasswordHash(PasswordUtil.hashPassword(dto.password()));
        portalUser.setRole(dto.role().trim().toLowerCase());
        portalUser.setCreatedAt(Instant.now());

        PortalUser saved = portalUserRepository.save(portalUser);
        return new PortalUserResponseDto(
                saved.getId(),
                saved.getName(),
                saved.getEmail(),
                saved.getRole(),
                saved.getCreatedAt().toString()
        );
    }
}

File: backend-nest-backup/src/admin-auth/admin-auth.service.ts
import {
  ConflictException,
  Injectable,
  UnauthorizedException,
} from '@nestjs/common';
import { PortalRole } from '@prisma/client';
import { createHash, randomUUID, timingSafeEqual } from 'crypto';

import { loadEnvFiles } from '../config/load-env';
import { PrismaService } from '../prisma/prisma.service';

import {
  AdminLoginDto,
  AdminSession,
  CreatePortalUserDto,
  PortalUserRecord,
} from './admin-auth.types';

interface TokenPayload {
  sub: string;
  email: string;
  name: string;
  role: AdminSession['role'];
  exp: number;
}

@Injectable()
export class AdminAuthService {
  private readonly tokenTtlMs = 1000 * 60 * 60 * 12;

  constructor(private readonly prisma: PrismaService) {}

  private hashPassword(password: string): string {
    return createHash('sha256').update(password).digest('hex');
  }

  private safeCompare(a: string, b: string): boolean {
    const left = Buffer.from(a, 'utf8');
    const right = Buffer.from(b, 'utf8');
    if (left.length !== right.length) return false;
    return timingSafeEqual(left, right);
  }

  private getJwtSecret(): string {
    const secret = process.env.ADMIN_JWT_SECRET ?? process.env.JWT_SECRET;
    if (!secret) {
      throw new UnauthorizedException('ADMIN_JWT_SECRET não configurado no .env');
    }
    return secret;
  }

  private readSuperadminEnv(): { email: string; password: string; name: string } | null {
    this.reloadEnvFromDisk();

    const email = process.env.SUPERADMIN_EMAIL?.trim().toLowerCase();
    const password = process.env.SUPERADMIN_PASSWORD?.replace(/\r$/, '') ?? '';
    if (!email || !password) return null;
    return {
      email,
      password,
      name: process.env.SUPERADMIN_NAME?.trim() || 'Super Admin',
    };
  }

  isSuperadminConfigured(): boolean {
    return this.readSuperadminEnv() !== null;
  }

  private reloadEnvFromDisk(): void {
    if (process.env.NODE_ENV === 'production') return;
    loadEnvFiles(true);
  }

  private signToken(payload: Omit<TokenPayload, 'exp'>): string {
    const fullPayload: TokenPayload = {
      ...payload,
      exp: Date.now() + this.tokenTtlMs,
    };
    const body = Buffer.from(JSON.stringify(fullPayload)).toString('base64url');
    const signature = createHash('sha256')
      .update(`${body}.${this.getJwtSecret()}`)
      .digest('base64url');
    return `${body}.${signature}`;
  }

  verifyToken(token: string): AdminSession {
    const [body, signature] = token.split('.');
    if (!body || !signature) {
      throw new UnauthorizedException('Token inválido.');
    }

    const expected = createHash('sha256')
      .update(`${body}.${this.getJwtSecret()}`)
      .digest('base64url');

    if (!this.safeCompare(signature, expected)) {
      throw new UnauthorizedException('Token inválido.');
    }

    const payload = JSON.parse(Buffer.from(body, 'base64url').toString('utf-8')) as TokenPayload;

    if (payload.exp < Date.now()) {
      throw new UnauthorizedException('Sessão expirada. Faça login novamente.');
    }

    return {
      id: payload.sub,
      email: payload.email,
      name: payload.name,
      role: payload.role,
    };
  }

  async login(dto: AdminLoginDto): Promise<{ token: string; user: AdminSession }> {
    const email = dto.email.trim().toLowerCase();
    const superadmin = this.readSuperadminEnv();

    if (superadmin && email === superadmin.email) {
      if (!this.safeCompare(dto.password, superadmin.password)) {
        throw new UnauthorizedException('E-mail ou senha inválidos.');
      }

      const user: AdminSession = {
        id: 'superadmin',
        name: superadmin.name,
        email: superadmin.email,
        role: 'superadmin',
      };

      return {
        user,
        token: this.signToken({
          sub: user.id,
          email: user.email,
          name: user.name,
          role: user.role,
        }),
      };
    }

    const portalUser = await this.prisma.portalUser.findUnique({ where: { email } });
    if (!portalUser || portalUser.passwordHash !== this.hashPassword(dto.password)) {
      throw new UnauthorizedException('E-mail ou senha inválidos.');
    }

    const user: AdminSession = {
      id: portalUser.id,
      name: portalUser.name,
      email: portalUser.email,
      role: portalUser.role,
    };

    return {
      user,
      token: this.signToken({
        sub: user.id,
        email: user.email,
        name: user.name,
        role: user.role,
      }),
    };
  }

  async listPortalUsers(): Promise<Omit<PortalUserRecord, 'passwordHash'>[]> {
    const users = await this.prisma.portalUser.findMany({
      orderBy: { createdAt: 'desc' },
    });

    return users.map((user) => ({
      id: user.id,
      name: user.name,
      email: user.email,
      role: user.role,
      createdAt: user.createdAt.toISOString(),
    }));
  }

  async createPortalUser(
    dto: CreatePortalUserDto,
  ): Promise<Omit<PortalUserRecord, 'passwordHash'>> {
    const email = dto.email.trim().toLowerCase();
    const superadmin = this.readSuperadminEnv();

    if (superadmin && superadmin.email === email) {
      throw new ConflictException('Este e-mail já está reservado ao superadmin.');
    }

    const existing = await this.prisma.portalUser.findUnique({ where: { email } });
    if (existing) {
      throw new ConflictException('Este e-mail já está cadastrado no portal.');
    }

    const record = await this.prisma.portalUser.create({
      data: {
        id: randomUUID(),
        name: dto.name.trim(),
        email,
        passwordHash: this.hashPassword(dto.password),
        role: dto.role as PortalRole,
      },
    });

    return {
      id: record.id,
      name: record.name,
      email: record.email,
      role: record.role,
      createdAt: record.createdAt.toISOString(),
    };
  }
}
